Add support for 3ds in change PM for subscription flow

timur27 opened the pull request:
Fixes https://github.com/Automattic/woocommerce-payments/issues/2822

#### Changes proposed in this Pull Request

This PR adds support for subscription payment method change with 3DS cards. It interferes with core subscriptions and hooks onto an action that allows us stop updating payment methods for subscriptions in order to be able to authenticate them first. 

### Testing instructions

#### Change payment method flow
* Purchase a subscription with a normal non-3DS card
* Navigate to change payment method flow and change payment method to a 3DS `4000002760003184` with the following use-cases
    * Upon authorization, cancel the modal window from the top right side and confirm 
        * The `We are unable to authenticate your payment method. Please choose a different payment method and try again` error on the page
        * Subscription has the old payment method still attached
    * Click on "Change payment method" once again, this time fail the authorization and confirm that
        * `We are unable to authenticate your payment method. Please choose a different payment method and try again.` notice is visible
        * Subscription has the old payment method still attached
    * Uncheck the `Use this payment method for all of my current subscriptions` checkbox and confirm the authorization
        * Once the confirmation modal pops up, confirm there is no success notice displayed behind the pop-up window since we're still in the process of confirming the payment for 3DS
        * Confirm that you are redirected to the subscription details page under my account
        * Confirm tha the successful `Payment method updated.` notice is displayed
    * Check the `Use this payment method for all of my current subscriptions` checkbox and confirm the authorization
        * Confirm that you are redirected to the subscription details page under my account
        * Confirm that the successful `Payment method updated for all your current subscriptions.` notice is displayed
        * Confirm that all the subscriptions were updated with the new payment method
* Change payment method to a new non-3DS card
* Change payment method to a saved card (both 3DS and non-3DS)
* Use the insufficient funds 3DS card `4000000000009995` and confirm the card is not attached to the subscription after failure, and the correct error notice for insufficient funds is being displayed
* Process renewal for a subscription that has a 3DS card attached to it as a result of a change payment method flow
    * Upon renewal processing attempt, the order should be expectedly put `On hold` and once you navigate to the link in the renewal order email sent to the shopper, you should be able to successfully renew order with a 3DS card
    * Perform the same test, but change to a non-3DS card during renewal order


-------------------

- [ ] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [ ] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
If this PR need not be QA tested, edit to 'QA Testing Not Applicable'
-->

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

botwoo commented:
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `10453` or branch name `add/change-pm-3ds-flow` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=add/change-pm-3ds-flow) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: e03f91646773f7c57685779cd22a4af4744d02ee
- Build time: 2025-03-11 07:54:41 UTC

_Note: the build is updated when a new commit is pushed to this PR._

timur27 committed:
tmp

timur27 committed:
attach intent info to order despite exitigin early for the flow

timur27 committed:
return formatting

timur27 committed:
let change payment method flow update the intent id for order

timur27 committed:
ensure correct return url for subscription pm change with 3ds

timur27 committed:
let the form template show the update-all checkbox

timur27 committed:
simplify non-3ds subscription update

timur27 committed:
make the 3ds flow a little bit more clear

timur27 committed:
slightly simplify the 3ds follow-up flow

timur27 committed:
fallback to default method in case subscriptions one is unavailable

timur27 committed:
fix linting and formatting

timur27 committed:
add changelog entry

timur27 committed:
Merge branch 'develop' of github.com:Automattic/woocommerce-payments into add/change-pm-3ds-flow

timur27 committed:
switch the checks

timur27 committed:
fix psalm error

timur27 commented on includes/class-wc-payment-gateway-wcpay.php:
```diff
@@ -1697,10 +1697,19 @@ public function process_payment_for_order( $cart, $payment_information, $schedul
 						'payment_method' => $payment_information->get_payment_method(),
 					];
 				}
+			} elseif ( $this->is_changing_payment_method_for_subscription() ) {
+				// Only attempt to use WC_Subscriptions_Change_Payment_Gateway if it exists.
+				if ( class_exists( 'WC_Subscriptions_Change_Payment_Gateway' ) ) {
+					// Update the payment method for subscription if the payment intent is not requiring action.
+					WC_Subscriptions_Change_Payment_Gateway::update_payment_method( $order, $payment_information->get_payment_method() );
+				}
+
+				// Because this new payment does not require action/confirmation, remove this filter so that WC_Subscriptions_Change_Payment_Gateway proceeds to update all subscriptions if flagged.
+				remove_filter( 'woocommerce_subscriptions_update_payment_via_pay_shortcode', [ $this, 'update_payment_method_for_subscriptions' ], 10 );
```
This is to manually update subscription with the new payment method which is non-3DS

timur27 commented on includes/class-wc-payment-gateway-wcpay.php:
```diff
@@ -1710,6 +1719,17 @@ public function process_payment_for_order( $cart, $payment_information, $schedul
 
 		$this->maybe_add_customer_notification_note( $order, $processing );
 
+		if ( isset( $status ) && Intent_Status::REQUIRES_ACTION === $status && $this->is_changing_payment_method_for_subscription() ) {
+			// Because we're filtering woocommerce_subscriptions_update_payment_via_pay_shortcode, we need to manually set this delayed update all flag here.
+			if ( isset( $_POST['update_all_subscriptions_payment_method'] ) && wc_clean( wp_unslash( $_POST['update_all_subscriptions_payment_method'] ) ) ) { // phpcs:ignore WordPress.Security.NonceVerification.Missing
+				$order->update_meta_data( '_delayed_update_payment_method_all', wc_clean( wp_unslash( $_POST['payment_method'] ) ) ); // phpcs:ignore WordPress.Security.NonceVerification.Missing
+				$order->save();
+			}
+
+			wp_safe_redirect( $response['redirect'] );
+			exit;
+		}
+
```
This is to set the stage for 3DS handling and force quit to stop subscriptions from making unwanted changes, as we will handle once 3DS is confirmed later by `update_order_status` AJAX.

timur27 commented on includes/class-wc-payment-gateway-wcpay.php:
```diff
@@ -3488,10 +3510,27 @@ public function update_order_status() {
 					}
 				}
 
+				$return_url = $this->get_return_url( $order );
+
+				if ( $is_changing_payment ) {
+					$payment_token = $this->get_payment_token( $order );
+					if ( class_exists( 'WC_Subscriptions_Change_Payment_Gateway' ) ) {
+						WC_Subscriptions_Change_Payment_Gateway::update_payment_method( $order, $payment_token->get_gateway_id() );
+						$notice = __( 'Payment method updated.', 'woocommerce-payments' );
+
+						if ( WC_Subscriptions_Change_Payment_Gateway::will_subscription_update_all_payment_methods( $order ) && WC_Subscriptions_Change_Payment_Gateway::update_all_payment_methods_from_subscription( $order, $token->get_gateway_id() ) ) {
+							$notice = __( 'Payment method updated for all your current subscriptions.', 'woocommerce-payments' );
+						}
+
+						wc_add_notice( $notice );
+					}
+					$return_url = method_exists( $order, 'get_view_order_url' ) ? $order->get_view_order_url() : $this->get_return_url( $order );
+				}
+
```
This is the follow up for 3DS that we prepared for in https://github.com/Automattic/woocommerce-payments/pull/10453/files#r1976071421

timur27 commented on includes/compat/subscriptions/trait-wc-payment-gateway-wcpay-subscriptions.php:
```diff
@@ -225,6 +225,34 @@ public function maybe_init_subscriptions_hooks() {
 
 		// Update subscriptions token when user sets a default payment method.
 		add_filter( 'woocommerce_subscriptions_update_subscription_token', [ $this, 'update_subscription_token' ], 10, 3 );
+		add_filter( 'woocommerce_subscriptions_update_payment_via_pay_shortcode', [ $this, 'update_payment_method_for_subscriptions' ], 10, 3 );
+	}
+
+	/**
+	 * Stops WC Subscriptions from updating the payment method for subscriptions.
+	 *
+	 * @param bool            $update_payment_method Whether to update the payment method.
+	 * @param string          $new_payment_method The new payment method.
+	 * @param WC_Subscription $subscription The subscription.
+	 * @return bool
+	 */
+	public function update_payment_method_for_subscriptions( $update_payment_method, $new_payment_method, $subscription ) {
+		// Skip if the change payment method request was not made yet.
+		if ( ! isset( $_POST['_wcsnonce'] ) || ! wp_verify_nonce( sanitize_key( $_POST['_wcsnonce'] ), 'wcs_change_payment_method' ) ) {
+			return $update_payment_method;
+		}
+
+		// Avoid interfering with use-cases not related to updating payment method for subscriptions.
+		if ( ! $this->is_changing_payment_method_for_subscription() ) {
+			return $update_payment_method;
+		}
+
+		// Avoid interfering with other payment gateways' operations.
+		if ( $new_payment_method !== $this->id ) {
+			return $update_payment_method;
+		}
+
+		return false;
```
This is the main hook that lets us stop the normal subscription payment method change flow executed by [core subscriptions](https://github.com/Automattic/woocommerce-subscriptions-core/blob/trunk/includes/class-wc-subscriptions-change-payment-gateway.php#L279)

timur27 commented on includes/class-wc-payments-order-service.php:
```diff
@@ -901,12 +901,13 @@ public function get_fraud_meta_box_type_for_order( $order ): string {
+	public function attach_intent_info_to_order( WC_Order $order, $intent, $is_changing_payment_method_for_subscription = false ) {
 		// We don't want to allow metadata for a successful payment to be disrupted.
-		if ( Intent_Status::SUCCEEDED === $this->get_intention_status_for_order( $order ) ) {
+		if ( Intent_Status::SUCCEEDED === $this->get_intention_status_for_order( $order ) && ! $is_changing_payment_method_for_subscription ) {
```
We should let the change payment method flow update the intent in order even though the previous intent is successful, that's why we are not letting this function to stop the flow for successful intents with change payment method use-case.

timur27 commented on includes/class-wc-payment-gateway-wcpay.php:
```diff
@@ -3488,10 +3510,27 @@ public function update_order_status() {
 				echo wp_json_encode(
 					[
-						'return_url' => $this->get_return_url( $order ),
+						'return_url' => $return_url,
```
The default `get_return_url` redirects users to the order received page because subscription change nonce is gone at this point as we're in the 3DS confirmation follow up AJAX here, that's why I'm using `get_view_order_url`.

timur27 commented on includes/compat/subscriptions/trait-wc-payment-gateway-wcpay-subscriptions.php:
```diff
@@ -225,6 +225,34 @@ public function maybe_init_subscriptions_hooks() {
+	 */
+	public function update_payment_method_for_subscriptions( $update_payment_method, $new_payment_method, $subscription ) {
+		// Skip if the change payment method request was not made yet.
+		if ( ! isset( $_POST['_wcsnonce'] ) || ! wp_verify_nonce( sanitize_key( $_POST['_wcsnonce'] ), 'wcs_change_payment_method' ) ) {
+			return $update_payment_method;
+		}
```
This is to let the "update payment method for all subscriptions" checkbox display, as the action this code hooks onto is being executed by [the change payment method form on the page load.](https://github.com/Automattic/woocommerce-subscriptions-core/blob/trunk/templates/checkout/form-change-payment-method.php#L66)

timur27 commented on includes/class-wc-payment-gateway-wcpay.php:
```diff
@@ -3437,8 +3457,10 @@ public function update_order_status() {
+			$this->is_changing_payment_method_for_subscription();
 
-			if ( $amount > 0 ) {
+			if ( $amount > 0 && ! $is_changing_payment ) {
```
Avoid entering the pi flow even though orders of type subscription might have amount > 0.

timur27 commented on includes/class-wc-payment-gateway-wcpay.php:
```diff
@@ -1697,10 +1697,19 @@ public function process_payment_for_order( $cart, $payment_information, $schedul
 						'payment_method' => $payment_information->get_payment_method(),
 					];
 				}
+			} elseif ( $this->is_changing_payment_method_for_subscription() ) {
```
I considered adding an extra filter to execute this code only for `succeeded` setup intent statuses, skipping `processing` and `requires_capture`. Besides the information gathered from Stripe docs that those statuses are unlikely (and for the `requires_capture` this status is never set for change payment method flow p1740995860215559-slack-C01BPL3ALGP), the main conclusion I made was that for non-3DS cards historically, our payment processing implementation resulted in a successful response for `PROCESSING` and `REQUIRES_CAPTURE` statuses, which for [the change_payment_method flow](https://github.com/Automattic/woocommerce-subscriptions-core/blob/trunk/includes/class-wc-subscriptions-change-payment-gateway.php#L279) meant that it still updated subscription status for those payments. 

Having said all that, I think entering this flow for statuses other than `REQUIRES_ACTION` makes sense. LMK how does it sound to you.

timur27 committed:
some cosmetic changes

timur27 committed:
support the subscription renewal use-case

timur27 committed:
add comments

timur27 committed:
add tests

james-allan commented:
### 🧪 Test cases

**3DS Payment method authorization failures**

- [x] Upon canceling the modal window from the top right side, confirm:
  - [x] "We are unable to authenticate your payment method. Please choose a different payment method and try again." error is displayed.
  - [x] Subscription still has the old payment method attached.
  - [x] Processing a renewal successfully used the old card.
- [x] Upon failing the authorization, confirm:
  - [x] "We are unable to authenticate your payment method. Please choose a different payment method and try again." notice is visible.
  - [x] Subscription still has the old payment method attached.
  - [x] Processing a renewal successfully used the old card.

**3DS Payment method authorization success - `4000002760003184`**

- [x] If the **"Use this payment method for all of my current subscriptions"** checkbox is **unchecked**, confirm:
  - [x] No success notice is displayed behind the confirmation modal window.
  - [x] You are redirected to the subscription details page under **My Account**.
  - [x] "Payment method updated." success notice is displayed.
- [x] If the **"Use this payment method for all of my current subscriptions"** checkbox is **checked**, confirm:
  - [x] You are redirected to the subscription details page under **My Account**.
  - [x] "Payment method updated for all your current subscriptions." success notice is displayed.
  - [x] All subscriptions were updated with the new payment method.

**Edge cases**

- [x] When using the **insufficient funds 3DS card (4000000000009995)**, confirm:
  - [x] The card is **not attached** to the subscription after failure.
  - [x] The correct error notice for insufficient funds is displayed.

**Subscription renewal with 3DS payment method**

- [x] When processing a renewal for a subscription with a always auth 3DS card attached (**`4000002760003184`**), confirm:
  - [x] The order is placed **On hold** upon renewal processing attempt.
  - [x] Upon navigating to the link in the renewal order email, the order can be successfully renewed with a 3DS card.
- [x] Perform the same renewal test, but change to a **non-3DS card** during renewal order.

james-allan commented:
@timur27 I ran into an error (`Error: No such payment_intent: 'seti_1Qz4Px2HReucZ98RTRxC0v6N'`) when attempting to successfully update a subscription. 

I'm not sure if I'm doing something wrong. I've uploaded this video to show what I was doing: 

https://github.com/user-attachments/assets/de47c6a8-e556-424b-8527-4f41a684e3e3

timur27 commented:
@james-allan thanks for your review and testing so far! Interestingly, I cannot reproduce this so far. I first thought maybe you're trying to test a subscription that was created prior to the bugfix (which is a perfectly valid scenario to test) but then I noticed the subscription start date from `4 minutes ago` which made me think that this is probably not the case. Also from your screen recording I can see that we are trying to search for a setup intent doing via a payment intent request, although [this flow should be avoided ](https://github.com/Automattic/woocommerce-payments/pull/10453/files#diff-bcfd303d507c33a27671935e724ee64d0df30ff2a18bb6ef4c94f28d48190124R3463) per some changes in this PR, so I'm not sure why it's entering it.

I'll try further to reproduce it. In the meantime, feel free to add any more details you might find helpful. Thanks!

https://github.com/user-attachments/assets/90182e72-a5fb-48a5-8990-5f28258ab135



UPDATE: @james-allan I was able to reproduce it on JN site where I use the WooPayments package based on my branch. This doesn't behave the same as my local environment. I think the isChangingPayment flag [is not added to the global config](https://github.com/Automattic/woocommerce-payments/blob/develop/includes/class-wc-payments-checkout.php#L256-L257) due to one of the check failing. I don't think it's `is_changing_payment_method_for_subscription` because I see the `change_payment_method` in the URL path so I guess this has something to do with `is_subscriptions_enabled`. I'll be looking further into it, feel free to drop any ideas regarding WooPayments <-> Subscriptions integration that might be missing in the setup, if you have some.

timur27 commented:
@brettshumaker, when you'll be testing this PR, could you please let us with @james-allan know how this specific scenario looks for you? This is the happy-path with 3DS, basically. Thanks!

timur27 commented:
@james-allan @brettshumaker [The failure here](https://github.com/Automattic/woocommerce-payments/pull/10453#issuecomment-2699186402) probably happens due to some subscription packages missing in WooPayments' production build. To not block testing/review, can you please review/test locally if your environment allows you to, while I'll have a look at the problem.

james-allan commented:
> I was able to reproduce it on JN site where I use the WooPayments package based on my branch.

Ah, I was thinking while reading your message that it could be related to the zip generated on JN. 

> I think the isChangingPayment flag [is not added to the global config](https://github.com/Automattic/woocommerce-payments/blob/develop/includes/class-wc-payments-checkout.php#L256-L257) due to one of the check failing.

Unless I've misunderstood, I don't think this is the issue I'm seeing on my JN site.

On the change payment method page I see this: 

<img width="231" alt="Screenshot 2025-03-06 at 12 17 35 pm" src="https://github.com/user-attachments/assets/df20512e-242f-4234-93ce-96e9a00449cc" /> 

When the modal pops up, it does change to "1" however, looking at how the `isChangingPayment` is used, I don't think that would cause an issue. 

<img width="296" alt="Screenshot 2025-03-06 at 12 18 37 pm" src="https://github.com/user-attachments/assets/915ddcd2-a1fc-45da-a243-949e64945a6e" />

timur27 committed:
put config global var within the confirmation Promise

timur27 committed:
map boolean to string for the request parser to catch up

timur27 commented:
@james-allan Thanks for your help by providing some useful details! The problem turned out to be the request parser was not supporting booleans and those should be converted to strings. It worked locally because the value from the back-end was mapped to a `"1"` whereas in JN it's `true` and I needed to [map it to string](https://github.com/Automattic/woocommerce-payments/pull/10453/commits/f9c665128dbfbb0d0c5702f3273624821cb077a5), similarly to how it's done in [other place](https://github.com/Automattic/woocommerce-payments/blob/develop/client/checkout/api/index.js#L249-L251).


@james-allan @brettshumaker this is now ready to be further tested and reviewed. Thanks!

timur27 committed:
Merge branch 'develop' into add/change-pm-3ds-flow

timur27 committed:
Merge branch 'develop' into add/change-pm-3ds-flow

timur27 committed:
Merge branch 'develop' into add/change-pm-3ds-flow

brettshumaker commented:
(Borrowing this handy checklist)

### 🧪 Test cases

**3DS Payment method authorization failures**

- [x] Upon authorization, cancel the modal window from the top right side and confirm:
  - [x] The We are unable to authenticate your payment method. Please choose a different payment method and try again error on the page
  - [x] Subscription has the old payment method still attached
- [x] Upon failing the authorization, confirm:
  - [x] We are unable to authenticate your payment method. Please choose a different payment method and try again. notice is visible
  - [x] Subscription has the old payment method still attached


**3DS Payment method authorization success - `4000002760003184`**

- [x] Uncheck the Use this payment method for all of my current subscriptions checkbox and confirm the authorization:
  - [x] Once the confirmation modal pops up, confirm there is no success notice displayed behind the pop-up window since we're still in the process of confirming the payment for 3DS
  - [x] Confirm that you are redirected to the subscription details page under my account
  - [x] Confirm tha the successful Payment method updated. notice is displayed
- [x] Check the Use this payment method for all of my current subscriptions checkbox and confirm the authorization:
  - [x] Confirm that you are redirected to the subscription details page under my account
  - [x] Confirm that the successful Payment method updated for all your current subscriptions. notice is displayed
  - [x] Confirm that all the subscriptions were updated with the new payment method

**Changing payment method**
- [x] Change payment method to a new non-3DS card
- [ ] Change payment method to a saved card (both 3DS and non-3DS)
Ran into an issue here - changing payment with a saved card (both 3DS and non-3DS) and leaving the "Use this payment method for all of my current subscriptions" box checked only updates the one subscription - not all of them.

https://github.com/user-attachments/assets/d1beadf9-152b-4ac1-b2bb-8724c2e49f17


**Edge cases**

- [x] When using the **insufficient funds 3DS card (4000000000009995)**, confirm:
  - [x] The card is **not attached** to the subscription after failure.
  - [x] The correct error notice for insufficient funds is displayed.

**Subscription renewal with 3DS payment method**

- [x] When processing a renewal for a subscription with a 3DS card attached, confirm:
  - [x] The order is placed **On hold** upon renewal processing attempt.
  - [x] Upon navigating to the link in the renewal order email, the order can be successfully renewed with a 3DS card.
- [ ] Perform the same renewal test, but change to a **non-3DS card** during renewal order.

I noticed that when I click the link in the renewal email for the 3ds card, the selected payment method isn't the card attached to the subscription, but rather the first saved card on my account. Is this the expected behavior? One time it also just changed the payment method to "Cards."
<img width="296" alt="Screenshot 2025-03-07 at 2 23 07 PM" src="https://github.com/user-attachments/assets/40eda774-c3b3-4c18-bddc-c1e0a840bc56" />

james-allan commented:
> I noticed that when I click the link in the renewal email for the 3ds card, the selected payment method isn't the card attached to the subscription, but rather the first saved card on my account. Is this the expected behavior?

Yeah I think that's expected, I may be wrong but I'm pretty sure the selected payment method is the account's default. So going to the My Account > Payment methods and setting a new default should change the automatically selected one. 

> One time it also just changed the payment method to "Cards."

This is a question more for you folks. I see WooPayments hooks onto `woocommerce_subscription_note_new_payment_method_title` to filter the name of the payment method, I'd guess there's an issue in there. 


@timur27 thanks for fixing that issue I reported earlier. I went through and retested my test cases. They all passed as expected. From a WC Subscriptions point of view this UX looks correct to me.

timur27 committed:
do not stop the default update flow for saves payment methods

timur27 commented:
@brettshumaker thanks for the review, addressing two points mentioned by you:

> Ran into an issue here - changing payment with a saved card (both 3DS and non-3DS) and leaving the "Use this payment method for all of my current subscriptions" box checked only updates the one subscription - not all of them.

Fixed in https://github.com/Automattic/woocommerce-payments/pull/10453/commits/d636a6d058b1d7204b5dd7b5e75ca754b02798a7

> I noticed that when I click the link in the renewal email for the 3ds card, the selected payment method isn't the card attached to the subscription, but rather the first saved card on my account. Is this the expected behavior? One time it also just changed the payment method to "Cards."

This happens because payment information object is created during a flow different from `process payment` request (instead, upload order status AJAX). I was able to reproduce it on `develop` too, with renewal payments. [I created an issue for this](https://github.com/Automattic/woocommerce-payments/issues/10522), where I also described in more details how to reproduce it. 

@brettshumaker can I ask to review the PR after the most recent changes addressing the problem with updating payment method for all subscriptions? Thanks!

brettshumaker reviewed (approved):
I went back through and retested the one failing test scenario I encountered and see that it is resolved. The other test scenarios seem to be working as expected as well. Thanks for creating that issue to track the problem with the "Cards" title.

timur27 committed:
Merge branch 'develop' into add/change-pm-3ds-flow
